loopy/codegen/tools.py

```diff
@@ -0,0 +1,188 @@
+            callkernel_index.append(i-1)
+        elif isinstance(sched_item, ReturnFromKernel):
+            callkernel_index.append(None)
+        else:
+            callkernel_index.append(callkernel_index[i-1])
+
+    # }}}
+
+    # {{{ computing has_barrier_within
+
+    # has_barrier_within[i]: 'True' if the i-th schedule item's is a begin block
+    # containing a barrier of if the i-th schedule item is a barrier.
+
+    has_barrier_within = []
+
+    for sched_idx, sched_item in enumerate(kernel.schedule):
+        if isinstance(sched_item, BeginBlockItem):
+            # TODO: calls to "gather_schedule_block" can be amortized
+            _, endblock_index = gather_schedule_block(kernel.schedule,
+                    sched_idx)
+            has_barrier_within.append(any(
+                    isinstance(kernel.schedule[i], Barrier)
+                    for i in range(sched_idx+1, endblock_index)))
+        elif isinstance(sched_item, Barrier):
+            has_barrier_within.append(True)
+        else:
+            has_barrier_within.append(False)
+    # }}}
+
+    return CodegenOperationCacheManager(kernel, active_inames, callkernel_index,
+            has_barrier_within)
+
+
+class CodegenOperationCacheManager:
+    """
+    Caches operations arising during the codegen pipeline.
+
+    .. attribute:: kernel
+
+        An instance of :class:`loopy.LoopKernel` the cache manager is tied to.
+
+    .. attribute:: active_inames
+
+        An instance of :class:`list`, with the i-th entry being a :class:`frozenset`
+        of active inames at the point just before i-th schedule item.
+
+    .. attribute:: callkernel_index
+
+        An instance of :class:`list`, with the i-th entry being the index of
+        :class:`loopy.schedule.CallKernel` containing the point just before i-th
+        schedule item.
+
+    .. attribute:: has_barrier_within
+
+        An instance of :class:`list`. The list's i-th entry is *True* if the i-th
+        schedule item is a :class:`loopy.schedule.BeginBlockItem` containing a
+        barrier or if the i-th schedule item is a :class:`loopy.schedule.Barrier`.
+
+    .. automethod:: with_kernel
+    .. automethod:: get_parallel_inames_in_a_callkernel
+    """
+    def __init__(self, kernel, active_inames, callkernel_index, has_barrier_within):
+        self.kernel = kernel
+        self.active_inames = active_inames
+        self.callkernel_index = callkernel_index
+        self.has_barrier_within = has_barrier_within
+
+    def with_kernel(self, kernel):
+        """
+        Returns a new instance of :class:`CodegenOperationCacheManager`
+        corresponding to *kernel* if the cached variables in *self* would
+        be valid for *kernel*, else returns *self*.
+        """
+        if ((self.kernel.instructions != kernel.instructions)
+                or (self.kernel.schedule != kernel.schedule)
+                or (self.kernel.iname_to_tags != kernel.iname_to_tags)):
```

Comment:
The tag sets could probably be filtered down to just the code-gen-relevant ones, though that might be overkill. Maybe add a comment to that effect?

Reply:
Made this a TODO. 2061eea9